sample_factory/algo/runners/runner.py

```diff
@@ -490,11 +495,48 @@ def _on_start(self):
         """Override this in a subclass to do something right when the experiment is started."""
         self.sampler.init()
 
-    def _setup_component_heartbeat(self):
-        # TODO!!!
-        pass
+    def _setup_component_heartbeat(self, component: HeartbeatStoppableEventLoopObject):
+        """
+        Groups components with heartbeat mechanism by type and records starting time.
+        When all components of the same type do not respond in the reporting timeframe, stops the run
+        """
+        component_type = type(component)
+        if component_type not in self.heartbeat_dict:
+            self.heartbeat_dict[component_type] = {}
+        type_dict = self.heartbeat_dict[component_type]
+        type_dict[component.object_id] = time.time()
+        component.heartbeat.connect(self._receive_heartbeat)
+
+    def _receive_heartbeat(self, component_type: type, component_id: str):
+        """
+        Record the time the most recent heartbeat was received
+        """
+        curr_time = time.time()
+        heartbeat_time = self.heartbeat_dict[component_type][component_id]
+        if curr_time - heartbeat_time > self.heartbeat_report_sec:
+            log.info(f"Heartbeat reconnected after {curr_time - heartbeat_time} sec from {component_id}")
+        self.heartbeat_dict[component_type][component_id] = curr_time
 
-    def _setup_component_termination(self, stop_signal: signal, component_to_stop: StoppableEventLoopObject):
+    def _check_heartbeat(self):
+        """
+        Reports components whose last heartbeat signal is longer than self.heartbeat_report_sec.
+        If all components of the same time fail, stop the run
+        """
+        curr_time = time.time()
+        log.info("Checking heartbeat")
+        for component_type, heartbeat_dict in self.heartbeat_dict.items():
+            num_components = len(heartbeat_dict)
+            num_stopped = 0
+            for component_id, heartbeat_time in heartbeat_dict.items():
+                if curr_time - heartbeat_time > self.heartbeat_report_sec:
+                    log.error(f"No heartbeat for {curr_time - heartbeat_time} sec from {component_id}")
```

Comment:
Can you make it such that there's a single message reporting all failed components, and only print it if there's at least one.

Like: No heartbeat for components: RolloutWorker-0 (300 seconds), ... etc.